File: Cpp/Populating Next Right Pointers in Each Node.cpp
#include<iostream>


using namespace std;

struct TreeLinkNode {
    int val;
    TreeLinkNode *left, *right, *next;
    TreeLinkNode(int x) : val(x), left(NULL), right(NULL), next(NULL) {}
};

void connectTree(TreeLinkNode *root,TreeLinkNode *lchild,TreeLinkNode *rchild)
    {
        if(lchild==NULL||rchild==NULL) return;
        lchild->next=rchild;
        if(root->next!=NULL) rchild->next=(root->next)->left;
        else rchild->next=NULL;
        connectTree(lchild, lchild->left, lchild->right);
        connectTree(rchild, rchild->left, rchild->right);
    }


void connect(TreeLinkNode *root)
    {
        if(root==NULL) return;
        connectTree(root, root->left, root->right);
    }

File: Python/Populating-Next-Right-Pointers-in-Each-Node.py
#!/usr/bin/env python

# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
        self.next = None

class Solution:
    # @param root, a tree node
    # @return nothing
    def findnextnode(self, root):
        if root.next == None:
            return None
        return root.next.left;

    def connect(self, root):
        if root == None:
            return
        if root.left == None:
            return
        root.left.next = root.right;
        root.right.next = self.findnextnode(root)
        self.connect(root.left);
        self.connect(root.right);